ethers-solc/src/config.rs

```diff
@@ -171,6 +171,61 @@ impl ProjectPathsConfig {
     pub fn find_libs(root: impl AsRef<Path>) -> Vec<PathBuf> {
         vec![utils::find_fave_or_alt_path(root, "lib", "node_modules")]
     }
+
+    /// Flattens all file imports into a single string
+    pub fn flatten(&self, target: &Path) -> Result<String> {
+        tracing::trace!("flattening file");
+        let graph = Graph::resolve(self)?;
+        self.flatten_node(target, &graph, false)
+    }
+
+    /// Flattens a single node from the dependency graph
+    fn flatten_node(
+        &self,
+        target: &Path,
+        graph: &Graph,
+        omit_version_pragma: bool,
+    ) -> Result<String> {
+        let target_dir = target.parent().ok_or_else(|| {
+            SolcError::msg(format!("failed to get parent directory for \"{:?}\"", target.display()))
+        })?;
+        let target_index = graph.files().get(target).ok_or_else(|| {
+            SolcError::msg(format!("cannot resolve file at \"{:?}\"", target.display()))
+        })?;
+        let target_node = graph.node(*target_index);
+
+        let mut imports = target_node.imports().clone();
+        imports.sort_by(|a, b| a.loc().0.cmp(&b.loc().0));
+
+        let mut content = target_node.content().bytes().collect::<Vec<_>>();
+        let mut offset = 0_isize;
+
+        if omit_version_pragma {
+            if let Some(version) = target_node.version() {
+                let (start, end) = version.loc();
+                content.splice(start..end, std::iter::empty());
+                offset -= (end - start) as isize;
+            }
+        }
+
+        for import in imports.iter() {
+            let import_path = self.resolve_import(target_dir, import.path())?;
+            let import_content = self.flatten_node(&import_path, graph, true)?;
+            let import_content = import_content.trim().as_bytes().to_owned();
+            let import_content_len = import_content.len() as isize;
+            let (start, end) =
+                import.loc_by_offset(offset).expect("failed to determine import location");
```

Comment:
I wonder if there's a way to make offset `usize`?
perhaps flip it and do `offset +=`?

The offset should keep track of the adjusted length.

Oh, I can see now why this can be negative.

But the cast to size in `loc_by_offset` which requires a Result is a bit redundant if we unwrap it.

can we do loc_by_offset directly here, because we're also casting here again
`let import_offset = import_content_len - ((end - start) as isize);`

there should also be `isize::saturated_sub()` and `saturated_add`